src/_P077_CSE7766.ino

```diff
@@ -165,40 +164,26 @@ boolean Plugin_077(byte function, struct EventStruct *event, String& string)
     case PLUGIN_SERIAL_IN:
       {
         if (Plugin_077_init) {
-          if (Serial.available() > 0) {
-            serial_in_byte = Serial.read();
-            success = true;
-
-            if (cse_receive_flag) {
-              serial_in_buffer[serial_in_byte_counter++] = serial_in_byte;
-              if (24 == serial_in_byte_counter) {
-                addLog(LOG_LEVEL_DEBUG_DEV, F("CSE: packet received"));
-                CseReceived(event);
-                cse_receive_flag = 0;
-                serial_in_byte_counter = 0;
-                // new packet received, update values
-                UserVar[event->BaseVarIndex] = energy_voltage;
-                UserVar[event->BaseVarIndex + 1] = energy_power;
-                UserVar[event->BaseVarIndex + 2] = energy_current;
-                UserVar[event->BaseVarIndex + 3] = cf_pulses;
-//                break;
-              }
-            } else {
-              if (0x5A == serial_in_byte) { // 0x5A - Packet header 2
-                cse_receive_flag = 1;
-                addLog(LOG_LEVEL_DEBUG_DEV, F("CSE: Header received"));
-              } else {
-                serial_in_byte_counter = 0;
-              }
-              serial_in_buffer[serial_in_byte_counter++] = serial_in_byte;
+          success = true;
+          /* ONLINE CHECKSUMMING by Bartlomiej Zimon*/
+          bool P077_found = false;
+          while (Serial.available()>0 && !P077_found){ // if we have data try to find good checksum
+            uint8_t serial_in_byte = Serial.read();
+            P077_checksum -= serial_in_buffer[2]; // remove from checksum data to be removed
+            memmove(serial_in_buffer,serial_in_buffer+1,sizeof(serial_in_buffer)-1);
+            serial_in_buffer[25] = serial_in_byte;
+            P077_checksum += serial_in_buffer[22]; // add online checksum
+            if (P077_checksum == serial_in_buffer[23] && serial_in_buffer[1] == 0x5A){
+              P077_found = true;
+              addLog(LOG_LEVEL_DEBUG, F("CSE: packet found"));
+              CseReceived(event);
```

Comment:
I guess calling this should be part of the read call, not the `PLUGIN_SERIAL_IN` case.

The `PLUGIN_SERIAL_IN` must be as short as possible and if it found something you could schedule another read command in the scheduler to call the actual read function, which will process the buffer.
The read call is the normal call which is expected to deliver a new value.

Reply:
this is only proof of concept, but it works flawlesly from first launch, my vote for move it into 1/10s. timer

Reply:
@michass is testing this in more long term, we will see tomorrow if everything is fine.
We could use this concept also for PMSx003 plugin.
If You want I can move it to PLUGIN_TEN_PER_SECOND

Reply:
@TD-er do You think we need to tune here something more?

Reply:
Have found - `addLog(LOG_LEVEL_DEBUG, F("CSE: packet found"));` does not show this message in log, not sure if it is bug?

Reply:
@uzi18 If this appears to work, I will merge it and later we must think of a proper way to do this buffering of data.
I already started something similar, a few months ago, but never finished it yet.
See [this branch](https://github.com/TD-er/ESPEasy/commits/feature/serial_reader) (commits from July 24 and further)

Reply:
My idea is to keep 1 pkt + header of next packet, P077_checksum has got sum of elements from 2-22, to check with element 23.
We can also add check if header of next pkt is also ok, so we are extra sure everything is fine

Reply:
37999: CSE: B0: 85+90+2+230+248+0+3+43
37999: CSE: B8: 0+64+31+0+124+86+80+78
37999: CSE: B16: 152+1+171+143+241+153+161+219
37999: CSE: B24: 85+90+0+0+0+0+0+0
37999: CSE: s: 219

Reply:
@TD-er I'm realy suprised - we operate on data buffers but there are no options to show these buffer for simple debugging

Reply:
@uzi18 That's why I prefer a central handler to process these standard actions like fetching data in a circular buffer.
So we have to do the same debugging over and over again.

Reply:
we loos lots of time for this debugging

Reply:
Just one remark from me: I would drop out of the reception loop if serial.avaikable< number of bytes to receive. Wait for all bytes to come in before you process them. This is less blocking.